src/server/accept.rs

```diff
@@ -89,83 +74,111 @@ where
+                result = accept => match result {
+                    Ok(accepted) => accepted,
+                    Err(error) => {
+                        eprintln!("error(listener): {}", error);
+                        drop(permit);
+                        continue;
+                    }
+                },
+                // A shutdown signal was received.
+                _ = shutdown_rx.changed() => {
+                    drop(permit);
+                    break 'accept match *shutdown_rx.borrow_and_update() {
+                        Some(false) => 0.into(),
+                        Some(true) | None => 1.into(),
+                    }
                 }
             }
         };
 
-        // Spawn a task to serve the connection.
-        connections.spawn({
-            // Clone acceptor so negotiation can happen in the connection task.
+        connections.push_back({
             let mut acceptor = acceptor.clone();
-
-            // Clone the watch receiver so we can shutdown the connection if a
-            // ctrl+c signal is sent to the process.
             let mut shutdown_rx = shutdown_rx.clone();
 
-            // Create an AppService to serve the connection.
-            let service = AppService::new(Arc::clone(&app), max_body_size);
-
-            async move {
-                let result = match acceptor.accept(stream).await {
-                    Err(error) => Err(error.into()),
-                    Ok(accepted) => {
-                        // Create a new HTTP/2 connection.
-                        #[cfg(feature = "http2")]
-                        let connection = conn::http2::Builder::new(TokioExecutor::new())
-                            .timer(TokioTimer::new())
-                            .serve_connection(IoStream::new(accepted), service);
-
-                        // Create a new HTTP/1.1 connection.
-                        #[cfg(all(feature = "http1", not(feature = "http2")))]
-                        let connection = conn::http1::Builder::new()
-                            .timer(TokioTimer::new())
-                            .serve_connection(IoStream::new(accepted), service)
-                            .with_upgrades();
-
-                        tokio::pin!(connection);
-
-                        // Serve the connection.
-                        tokio::select! {
-                            result = &mut connection => result.map_err(|e| e.into()),
-                            _ = shutdown_rx.changed() => {
-                                connection.as_mut().graceful_shutdown();
-                                connection.await.map_err(|e| e.into())
-                            }
-                        }
+            // TODO: Attempt to get the task size in the tls example <= 1024.
+            task::spawn(async move {
+                let io = match acceptor.accept(stream).await {
+                    Ok(accepted) => TokioIo::new(accepted),
```

Comment:
We probably want to wrap the socket as late as possible here.

I understand the important of needing writes to the socket to be as fast as possible.

Therefore, I'm thinking we stick with `TokioIo` instead of `IoStream` and consider boxing the connection rather than the task or the socket. It should result in a predictable task size that is small enough to offer greater scalability but not so compact that we potentially overflow the stack.

Reply:
The correct approach here may vary based on the runtime that is used.